test/e2e_node/services/kubelet.go

```diff
@@ -193,18 +200,44 @@ func (e *E2EServices) startKubelet() (*server, error) {
 		return nil, err
 	}
 
-	cmdArgs = append(cmdArgs,
+	kubeArgs = append(kubeArgs,
 		"--network-plugin=kubenet",
 		"--cni-bin-dir", cniBinDir,
 		"--cni-conf-dir", cniConfDir)
 
 	// Keep hostname override for convenience.
 	if framework.TestContext.NodeName != "" { // If node name is specified, set hostname override.
-		cmdArgs = append(cmdArgs, "--hostname-override", framework.TestContext.NodeName)
+		kubeArgs = append(kubeArgs, "--hostname-override", framework.TestContext.NodeName)
 	}
 
 	// Override the default kubelet flags.
-	cmdArgs = append(cmdArgs, kubeletArgs...)
+	kubeArgs = append(kubeArgs, kubeletArgs...)
+
+	// If the config file feature gate is enabled, generate the file and remove the flags it applies to
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletConfigFile) {
+		kc, other, err := splitKubeletConfigArgs(kubeArgs)
+		if err != nil {
+			return nil, err
+		}
+		// replace kubeArgs with the new command line, which has had the KubeletConfiguration flags removed
+		kubeArgs = other
+		path, err := writeKubeletConfigFile(kc)
+		if err != nil {
+			return nil, err
+		}
+		// add the feature gates back to the command line, some (e.g. KubeletConfigFile) may be relevant before the file is loaded
```

Comment:
Why don't we use `framework.TestContext.FeatureGates`?

It seems that we only care about global feature gates here, instead of kubelet specific feature gates specified in `kubelet-flags`.

Reply:
offline chat: decided to simplify down to `framework.TestContext.FeatureGates`. It's a little safer from a dynamic config perspective to keep this command-line minimal.

Reply:
done